pkl-core/src/main/java/org/pkl/core/messaging/AbstractMessagePackEncoder.java

```diff
@@ -0,0 +1,184 @@
+/**
+ * Copyright © 2024 Apple Inc. and the Pkl project authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.pkl.core.messaging;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Collection;
+import java.util.Map;
+import java.util.function.Function;
+import org.msgpack.core.MessagePack;
+import org.msgpack.core.MessagePacker;
+import org.pkl.core.util.Nullable;
+
+public abstract class AbstractMessagePackEncoder implements MessageEncoder {
+
+  protected final MessagePacker packer;
+
+  public AbstractMessagePackEncoder(MessagePacker packer) {
+    this.packer = packer;
+  }
+
+  public AbstractMessagePackEncoder(OutputStream stream) {
+    this(MessagePack.newDefaultPacker(stream));
+  }
+
+  protected abstract @Nullable void encodeMessage(Message msg)
+      throws ProtocolException, IOException;
+
+  @Override
+  public final void encode(Message msg) throws IOException, ProtocolException {
+    packer.packArrayHeader(2);
+    packer.packInt(msg.type().getCode());
+    encodeMessage(msg);
+    packer.flush();
+  }
+
+  protected void packMapHeader(int size, @Nullable Object value1) throws IOException {
+    packer.packMapHeader(size + (value1 != null ? 1 : 0));
+  }
+
+  protected void packMapHeader(int size, @Nullable Object value1, @Nullable Object value2)
+      throws IOException {
+    packer.packMapHeader(size + (value1 != null ? 1 : 0) + (value2 != null ? 1 : 0));
+  }
+
+  protected void packMapHeader(
```

Comment:
I though about this, but figured this approach (which matches the original Kotlin implementation) might have been chosen for efficiency reasons. This code is in a pretty hot path, so if the goal is to avoid allocations or iteration that would make sense to me. Happy to change this if there's consensus!

Reply:
I don't think performance should be a concern here. The JVM can easily optimize varargs parameters, specially on a hot path.
On the other hand I don't have strong opinions here, it was more of a nit.

Reply:
I don't remember why the Kotlin side was written that way.

I think varargs would be slower than this because it would require a for loop (unless the JVM's JIT does some unrolling magic), but I don't think we'd have any perf issues even with a for loop.

Reply:
Let's just keep this; no need to rewrite something that works and is likely more performant.